File: 0008 String to Integer/StringtoInteger.java
class Solution {
    public int myAtoi(String s) {
        long ret = 0;
        boolean  isReading = false;
        boolean  isPostive = true;

        for (int i = 0; i < s.length(); i++)
        {
            char c = s.charAt(i);
            if (c == ' ')
            {
                if (isReading) break;
            }
            else if (c == '+')
            {
                if (isReading) break;
                else isReading = true;
            }
            else if (c == '-')
            {
                if (isReading) break;
                else
                {
                    isReading = true;
                    isPostive = false;
                }
            }
            else if ('0' <= c && c <= '9')
            {
                isReading = true;

                ret *= 10;
                if (isPostive) ret += c - '0';
                else ret -= c - '0';

                if (ret < Integer.MIN_VALUE) return Integer.MIN_VALUE;
                else if (Integer.MAX_VALUE < ret) return Integer.MAX_VALUE;
            }
            else break;
        }
        return (int)ret;
    }
}

File: 0008 String to Integer/StringtoInteger.cpp
class Solution {
public:
    int myAtoi(string str) {
        char *s = &str[0];
        long ret = 0;
        bool  isReading = false;
        bool  isPostive = true;

        while (*s) {
            if (*s == ' ') {
                if (isReading) {
                    return ret;
                }
            } else if (*s == '+') {
                if (isReading) {
                    return ret;
                } else {
                    isReading = 1;
                }
            } else if (*s == '-') {
                if (isReading) {
                    return ret;
                } else {
                    isReading = 1;
                    isPostive = 0;
                }
            }  else if ('0' <= *s && *s <= '9') {
                isReading = 1;

                ret *= 10;
                if (isPostive) {
                    ret += *s - '0';
                } else {
                    ret -= *s - '0';
                }

                if (ret < INT_MIN) {
                    return INT_MIN;
                } else if (INT_MAX < ret) {
                    return INT_MAX;
                }
            } else {
                return ret;
            }
            *s++;
        }
        return ret;
    }
};